router/dataplane_concurrency_model.gobra

```diff
@@ -0,0 +1,254 @@
+decreases n
+func MultiUpdateElemWitness2(ghost t io.Place, n, m int, ghost k Key, ghost s io.IO_dp3s_state_local, ghost y SharedArg) {
+  if n > 0 {
+    unfold MultiReadBio(t,n)
+    val := io.dp3s_iospec_bio3s_recv_R(t)
+    next := io.dp3s_iospec_bio3s_recv_T(t)
+    if val.isIO_val_Pkt2{
+      UpdateElemWitness(s.ibuf, y.IBufY, val.IO_val_Pkt2_1, val.IO_val_Pkt2_2)
+    }
+    MultiUpdateElemWitness2(next, n-1, m-1, k, add_ibuf(s, val), y)
+    fold MultiReadBio(t,n)
+  } 
+  
+  if 0 <= m && m <= n && MutliReadBio_correctIfs(t,n,m,k) {
+    fold MultiElemWitness(y.IBufY, k, MutliReadBio_IO_val(t,n,m))
+  }
+}
+
+pred MultiElemWitness(ghost y ElemRA, ghost k Key, ghost es seq[io.IO_val]) { 
+  // cannot be quantified because there might be multiple witnesses for the same element
+  len(es) > 0 ==> match es[0] {
+    case io.IO_val_Pkt2{_, _}:
+      ElemWitness(y, k, es[0].IO_val_Pkt2_2)
+    default: 
+      true
+  } && MultiElemWitness(y, k, es[1:])
+}
+
+pred MultiElemWitness0(ghost y ElemRA, ghost k Key, ghost es seq[io.IO_val], i int) { 
+  0 <= i && i < len(es) ==> match es[i] {
+    case io.IO_val_Pkt2{_, _}:
+       ElemWitness(y, k, es[i].IO_val_Pkt2_2)
+    default: 
+      true
+  } && MultiElemWitness0(y, k, es, i+1)
+}
+
+ghost 
+decreases 
+requires MultiElemWitness(y,k,es)
+ensures  MultiElemWitness0(y,k,es,0)
+func MultiElemWitnessConv(ghost y ElemRA, ghost k Key, ghost es seq[io.IO_val]) {
+  multiElemWitnessConvAux(y,k,es,0)
+}
+
+ghost
+requires i >= 0 
+requires MultiElemWitness(y,k,es[i:])
+ensures  MultiElemWitness0(y,k,es,i)
+decreases len(es)-i
+func multiElemWitnessConvAux(ghost y ElemRA, ghost k Key, ghost es seq[io.IO_val], i int) {
+  unfold MultiElemWitness(y,k,es[i:])
+  if i < len(es) {
+    multiElemWitnessConvAux(y,k,es,i+1)
+  }
+  fold MultiElemWitness0(y,k,es,i)
+}
+
+/////////////////////////////////////// RA definitions ///////////////////////////////////////
+type Key = option[io.IO_ifs]
+type Val = set[io.IO_pkt3]
+type Elem = io.IO_pkt3
+
+type ElemRA domain{} 
+
+pred ElemAuth(ghost m dict[Key]Val, ghost y ElemRA) 
+
+pred ElemWitness(ghost y ElemRA, ghost k Key, ghost e Elem) 
+
+ghost
+decreases 
+requires ElemAuth(m, y) && ElemWitness(y, k, e)
+ensures  ElemAuth(m, y) && ElemWitness(y, k, e) && k in domain(m) && e in AsSet(m[k])
+func ApplyElemWitness(ghost m dict[Key]Val, ghost y ElemRA, ghost k Key, ghost e Elem) 
+
+ghost 
+decreases
+requires ElemAuth(m, y) 
+ensures  ElemAuth(io.insert(m, k, e), y) && ElemWitness(y, k, e)
+func UpdateElemWitness(ghost m dict[Key]Val, ghost y ElemRA, ghost k Key, ghost e Elem) 
+
+ghost
+decreases
+ensures ElemAuth(m, y)
+func InitElemAuth(ghost m dict[Key]Val) (ghost y ElemRA) 
+
+
+/////////////////////////////////////// Utility ///////////////////////////////////////
+
+ghost 
+decreases
+pure func AsSet(s set[Elem]) set[Elem] {
+	return s
+}
```

Comment:
why do we have this here? Can we drop it?

Reply:
Unfortunately not. Gobra still has the issue that the internal datatypes don't match.

Reply:
can you please add a comment here explaining why this is here for future reference?

Reply:
I already did that. It should be visible in the newest version.